Automatically start monitor after upload

Tasssadar opened the pull request:
This makes the upload action immediately start serial monitor. This solves the problem with serial monitor preventing the upload action from working, and vice-versa. I was unable to figure out a better solution, given that vscode has no APIs to detect task completition, so the method used in Atom is not possible here :/

ivankravets commented on src/main.js:
```diff
@@ -117,6 +137,19 @@ class PlatformIOVSCodeExtension {
+  terminateUploadTask() {
+    if(this.inUploadTask) {
+      this.inUploadTask = false;
+      return vscode.commands.executeCommand("workbench.action.tasks.terminate").then(() => {
```
Does it terminate all process from opened Terminals?

Tasssadar replied:
It terminates currently running foreground task, of which there can only be one I think. The "Upload and Monitor" task keeps running in the foreground as long as the serial port is connected.

ivankravets commented:
>  I was unable to figure out a better solution, given that vscode has no APIs to detect task completition, so the method used in Atom is not possible here :/

I requested this API in https://github.com/Microsoft/vscode/issues/29855

Tasssadar commented:
Ok, I see you wanna make it work like in Atom, unfortunately we'd like to have this next week on our robotics camp to make it a bit less annoying for the kids. We should be able to use our private fork if you don't wanna merge this in the meantime until MS makes that API though.

ivankravets commented on src/main.js:
```diff
@@ -117,6 +137,19 @@ class PlatformIOVSCodeExtension {
     ));
   }
 
+  terminateUploadTask() {
```
Please rewrite this block with `async/await`
```js
  async terminateUploadTask() {
    if(!this._inUploadTask) {
return;
}
      this._inUploadTask = false;
try {
      await vscode.commands.executeCommand("workbench.action.tasks.terminate")
} catch (err) {
console.error(err);
}
        return new Promise((resolve, reject) => {
          setTimeout(() => resolve(), 100);
        });
      });
  }
```

P.S: I didn't test this code. Please valiate.

ivankravets commented on src/main.js:
```diff
@@ -19,6 +19,7 @@ class PlatformIOVSCodeExtension {
   constructor() {
     this._context = null;
     this.pioTerm = new PIOTerminal();
+    this.inUploadTask = false;
```
Please make a private variable `_inUploadTask`

ivankravets commented:
> Ok, I see you wanna make it work like in Atom, unfortunately we'd like to have this next week on our robotics

I will merge your PR and release new version today. I've mentioned VSCode PR to show that we also wait for that.

/cc @dbaeumer

Tasssadar commented:
Updated. Sorry it took so long, wasn't near a PC for a bit. BTW, I also removed the `closeSerialMonitor()` func, I think this PR replaces it and it didn't trigger in some cases due to `_lastCommand` being nulled by the `new()` func anyway.

ivankravets commented:
Thank you very much!

P.S: Could you say a few words about your workshop? Links?

ivankravets commented:
I significantly improve this PR/Issue. See https://github.com/platformio/platformio-vscode-ide/commit/1ce223ad9cd5f5469f4670f275371516a95a2d62

By default, PIO doesn't handle "Upload" comman as "Upload and Monitor". It will confuse users. If you need that, you can force "Upload and Monitor" with a new configuration option.

Tasssadar commented:
Ok, thanks!

We're RobotikaBrno -  Czech organization that does clubs and courses for kids during the school year about programming and electronics.

Every summer holidays, we're doing a week long camp where we teach kids and young adults (ages 10 to 20). Two years ago we used mostly LEGO EV3 with that hacked linux distro, last year we had Arduinos, and now we're gonna use ESP32 boards, via Visual Studio Code & the PlatformIO extension.

Here's a bunch of photos from last year: https://goo.gl/photos/rg6uC1f8ME6YJDPL7

ivankravets commented:
https://twitter.com/PlatformIO_Org/status/882662036785836036

ivankravets commented:
> now we're gonna use ESP32 boards, via Visual Studio Code & the PlatformIO extension.

You have done a greate choice! :) 

Happy coding with PlatformIO!

P.S: v0.4.0 is out! Please re-test. Thanks!

yaqwsx commented:
Hi,
I use this feature and it worked great. However, I think it started with VSCode 1.14, when I try to upload I always get "The task '{0}' is already active. To terminate it use `Terminate Task...` from the Tasks menu." I have to issue the upload command multiple times to actually start the upload. I think terminating the already running Serial monitor is broken. Could you look at it, please?

ivankravets commented:
@yaqwsx how to reproduce that?

1. Click on "Serial Monitor" icon on toolbar
2. Click on "Upload" icon on toolbar.
3. IDE asks to terminate process, we agree (temporary issue)
4. Uploading success.

P.S: Currently, we depend on https://github.com/Microsoft/vscode/issues/29855 but it seems that we will drop "Serial Monitor" from Terminal and move it to PIO Home.

yaqwsx commented:
When I have `"platformio-ide.forceUploadAndMonitor": false`, it works as you describe. However, when this option is true, I get following:

1. Run VSCode
2. Upload - program is compiled, uploaded and serial monitor is started
3. Press upload again
4. Agree to terminate task
5. Get  error: "The task '{0}' is already active. To terminate it use `Terminate Task...` from the Tasks menu."
6. Try upload again
7. Agree to terminate task
8. Get error: "Terminal with ID 2 does not exist (has it already been disposed?)"

ivankravets commented:
@yaqwsx  Please open new issue https://github.com/platformio/platformio-vscode-ide/issues

Thanks!
